Replace Nix tooling & update toolchain

sellout opened the pull request:
This replaces haskell.nix with stacklock2nix, and brings everything up to Nixpkgs 25.11 and Stackage LTS 24.21 (GHC 9.10.3).

It also adds documentation for how to upgrade tooling in future.

This is an alternative to #6033 + #5914.

## Implementation approach and notes

The first commit switches the infrastructure from haskell.nix to stacklock2nix. The latter is much simpler, but less flexible. We took advantage of none of the flexibility of haskell.nix (nor do I think there’s any reason to), but we had a lot of overhead and complexity due to it.

However, it’s difficult to switch the infrastructure without also updating the other tooling (Stack, GHC, etc.), so the remainder of the PR is largely about supporting the new tooling, mostly the upgrade from GHC 9.6.5 to 9.10.3.

## Loose ends

`x-partial` is a newly-introduced custom warning that has been applied to `head`, etc. I temporarily downgraded it from an error (because we use `-Werror`) to a warning. I have changes that allow us to restore it to an error again, but I think it’s better as a separate PR because it requires threading a lot of changes.

`unison-runtime:optchecks` is failing. @dolio is tracking that at https://github.com/nomeata/inspection-testing/issues/100.

sellout committed:
Document toolchain update process

sellout committed:
Switch from haskell.nix to stacklock2nix

This is much simpler Haskell tooling that serves our needs and keeps us
on a better cache (NixOS).

sellout committed:
Changes to support LFS 24.21 (GHC 9.10.3)

- resolve various implicit import conflicts (`show`, `unsnoc`, `unzip`, etc.)
- update `text-builder` names (`Text.Builder` -> `TextBuilder`, `run` ->
  `toText`)
- update primitive usage (`keepAlive` and
  `withMutableByteArrayContents`)

sellout committed:
Migrate to current lsp package

sellout committed:
Add Hoogle server to devShells

sellout committed:
De-escalate new warnings

The new `-Wx-partial` warning and some deprecations require more
involved changes, so defer for now.

sellout committed:
Remove unused imports

sellout committed:
Apply new Ormolu

sellout committed:
Update Stack version in GitHub workflows

sellout committed:
Add additional `allow-newer-deps` entries

sellout committed:
Bring cabal.project in sync with stack.yaml

sellout committed:
Update more versions scattered through the repo

sellout committed:
Update Weeder config

sellout committed:
Make the Weeder script more robust

There was a comment about this before, but the failure caused by having
both GHC 9.10.3 and 9.6.5 HIE files in the GitHub actions cache made it
worth addressing now. This also works around another bug introduced in
Weeder 1.9.0.

sellout committed:
Lift some file creation toward `main`

Deep file creation was causing tests to modify files in `$HOME`, which was exposed by Nix sandboxing. This moves the
file creation out a bit so that test code can specify an alternative location (alongside the test codebase).

I’m not sure this is the right level – should we just not manage credentials during a test? Have a null manager? And
this really should be moved up even further.

sellout commented:
@ChrisPenner This PR upgrades us to GHC 9.10 … and I think the compiler must be doing something differently somewhere, because the `optchecks` stuff is failing with effectively no change to `unison-runtime`. Maybe it just needs more of that `throwIO` workaround or something, but thought you might have a clue to offer.

sellout committed:
Regenerate all hpack files

__NB__: The hpack binary in Nixpkgs is for 0.38.3, but the library
        Stack is built against is 0.38.1 for some reason.

sellout committed:
Re-enable the Nix dev cache job on Linux

Also enable it on aarch64-linux for the first time.

Fixes #5789.

aryairani commented:
> `unison-runtime:optchecks` is failing, and I’m not sure why. Would love some assistance with this one.

@dolio If you might have an idea on this; possibly due to ghc version change?

dolio commented:
That test checks that GHC is generating code that looks properly optimized. I think it basically checks that the `eval` function contains no mention of the `Stack` type.

So, the details of the optimization have changed. I guess you could try running benchmarks before and after to see if the difference actually matters. If it doesn't, then maybe we need a different test. But if it does, we need to figure out how to fix the performance. Even if it doesn't, it might be easier to tweak something to satisfy the check than to figure out a new check that's actually proper.

sellout commented:
Here’s the Nix workflow run on my fork: https://github.com/sellout/unison/actions/runs/20785608716 (it’ll run on trunk in this repo when this PR is merged).

ceedubs reviewed (commented):
I think that the docs for building with cabal need to either switch back to `project-file` or need to otherwise give a working command (perhaps moving cabal files out of `contrib` or suggest adding a symlink). Otherwise this is working well for me.

sellout commented:
> I think that the docs for building with cabal need to either switch back to `project-file` or need to otherwise give a working command (perhaps moving cabal files out of `contrib` or suggest adding a symlink). Otherwise this is working well for me.

I removed the commit with the Cabal doc changes. I prefer the shorter `--project-dir=contrib` over `--project-file=contrib/cabal.project`, and I like the paths to be relative to the file they’re in … but it breaks the `ln -s contrib/* ./` workaround, which is nice for people who are using Cabal regularly.

sellout commented:
**NB**: An annoying thing here is that because of the way the Cabal workflow works (in order to allow forks to make comments on PRs), it’s going to fail until after this is merged – it uses the workflow file from the target branch (trunk), and that means it gets the old GHC (& Cabal) version.

ceedubs reviewed (approved):
Can't claim that I've reviewed the Haskell changes, but the Nix changes seem good to me :+1:

aryairani commented:
Ok but the current build failures aren't nix related right?
They're something weird about the optimization checks that needs to get fixed for ghc 9.10.3; so not currently mergeable.

sellout commented:
@dolio I tried running the benchmarks, but they don’t seem to build
```
⚙️   Processing stanza 1 of 15.
  ❓

    An error occurred while running the following file: unison-src/transcripts-manual/benchmarks.md

    ``` ucm :hide
    > pull unison.public.base.releases.M4d base
    > pull runarorama.public.sort.data sort
    ```

    🛑

    The transcript failed due to an error in the stanza above. The error is:

    ```
    I think you want to merge unison.public.base.releases.M4d into
    the base branch, but it doesn't exist. If you want, you can
    create it with `branch.empty base`, and then `pull` again.
    ```
```
At first I thought I was just doing something wrong, but I’ve had the same thing happen on various releases. Granted, it’s still possibly I’m doing something wrong. I‘m now seeing if it works for me at the point the benchmarks were last updated. If not, then it’s _very_ likely I’m doing something wrong.

Either way, the benchmarks should probably be run as part of CI. We don’t need to actually _measure_ them in CI (GitHub is no good for that), but at least ensure they build, so then I’d immediately know what I’m doing wrong.

aryairani commented:
> `> pull unison.public.base.releases.M4d base`

This is ancient syntax, so those aren't the right benchmarks. Those should be updated or deleted.

@dolio what are the right benchmarks that @sellout should run? Is `unison-src/transcripts-manual/benchmarks.md` a thing? Or, can you coordinate with them, thanks.

> That test checks that GHC is generating code that looks properly optimized. I think it basically checks that the eval function contains no mention of the Stack type.
> 
> So, the details of the optimization have changed. I guess you could try running benchmarks before and after to see if the difference actually matters. If it doesn't, then maybe we need a different test. But if it does, we need to figure out how to fix the performance. Even if it doesn't, it might be easier to tweak something to satisfy the check than to figure out a new check that's actually proper.

dolio commented:
The benchmarks are on share

https://share.unison-lang.org/@pchiusano/misc-benchmarks

run `suite` after pulling.

I don't actually remember what is in the benchmark transcript.

sellout commented:
Here are the benchmarks (times in µs). With one nasty exception, things look good. I re-ran them to see if there was some interference with that outlier, but everything looked the same.

  | GHC 9.6.5 | GHC 9.10.3 | change | benchmark
:-: | -: | -: | -: | :-
✔️ | 110.726 | 108.530 | -1.98% | Mutate a Ref 1000 times
✔️ | 45 410.000 | 43 397.000 | -4.43% | Mutate a local Remote.Ref 10k times
✔️ | 90.020 | 85.898 | -4.58% | Do 10k arithmetic operations
✔️ | 141.426 | 138.903 | -1.78% | List.map increment (range 0 1000)
✔️ | 578.737 | 391.878 | -32.29% | List.map murmurHash (range 0 1000)
✔️ | 70.645 | 62.998 | -10.82% | Multimap.fromList (range 0 1000)
✔️ | 5 767.393 | 5 603.962 | -2.83% | Stream functions
❌ | 3 265.559 | 10 221.946 | 213.02% | Value.serializeUncompressed (10k element map)
✔️ | 23 243.425 | 18 514.487 | -20.35% | Value.serializeCompressed (10k element map)
✔️ | 27 514.550 | 21 939.512 | -20.26% | Value.deserializeCompressed (10k element map)
✔️ | 8.914 | 7.961 | -10.69% | Json.toText (per document)
✔️ | 9.708 | 7.127 | -26.59% | Json parsing (per document)
✔️ | 14.671 | 10.520 | -28.29% | Json complex parsing (per document)
❌ | 61.385 | 62.808 | 2.32% | Json complex decoding (per document)
✔️ | 0.174 | 0.170 | -2.30% | Decode Nat
✔️ | 84.720 | 82.710 | -2.37% | Generate 100 random numbers
❌ | 1 169.958 | 1 171.398 | 0.12% | List.foldLeft
✔️ | 43 123.512 | 42 159.812 | -2.23% | Count to 1 million
  | 0.079 | 0.079 | 0.00% | Count to N (per element)
✔️ | 79.544 | 79.209 | -0.42% | Count to 1000
✔️ | 125.211 | 122.582 | -2.10% | CAS an IO.ref 1000 times
✔️ | 0.045 | 0.042 | -6.67% | List.range (per element)
✔️ | 61.784 | 55.654 | -9.92% | List.range 0 1000
✔️ | 29.812 | 29.658 | -0.52% | Set.fromList (range 0 1000)
✔️ | 28.705 | 28.232 | -1.65% | Map.fromList (range 0 1000)
✔️ | 2 660.596 | 2 603.665 | -2.14% | NatMap.fromList (range 0 1000)
✔️ | 0.190 | 0.178 | -6.32% | Map.lookup (1k element map)
✔️ | 0.265 | 0.254 | -4.15% | Map.insert (1k element map)
✔️ | 1 352.450 | 1 311.557 | -3.02% | Shuffle a 1000 element array
✔️ | 3 135.777 | 3 097.065 | -1.23% | Mutably mergesort a 1000 element array
✔️ | 0.145 | 0.143 | -1.38% | List.at (1k element list)
❌ | 2.742 | 2.803 | 2.22% | Text.split /
✔️ | 5 720.393 | 5 496.593 | -3.91% | Two match
✔️ | 5 905.034 | 5 668.678 | -4.00% | Four match
✔️ | 5 344.256 | 5 183.918 | -3.00% | Thirty match
✔️ | 137.770 | 133.497 | -3.10% | fib1
✔️ | 327.592 | 304.789 | -6.96% | fib2
✔️ | 331.159 | 303.199 | -8.44% | fib3

ceedubs commented:
> ❌ 	3 265.559 	10 221.946 	213.02% 	Value.serializeUncompressed (10k element map)

For what it's worth, this benchmark looks to be using `Value.serialize_v4`. Most things would be using v5 (`Value.serialize`) these days, so that would probably be the more relevant benchmark. It probably isn't a big deal if v4 serialization gets twice as slow. Though it's also strange because the compressed version of the test still using `serialize_v4` and then just sends it through zlib compression/decompression. That makes it seem like either a fluke or that there was a boost in the zlib performance that is offsetting a performance regression in the v4 serialization.

aryairani commented:
loose ends after updating stack is to add the stack.yaml flag to be lenient about hpack versions

pchiusano commented:
@dolio and I just talked about this:

* Dan's going to see if he can figure out why the Stack optimization check is firing. This will be timeboxed, if he can't figure it out in 1.5 days, we'll disable the check for now and rely on benchmarks to make sure we don't break performance if we make changes to the runtime before we get around to fixing the check.
* @sellout could you get the PR otherwise all ready to go?

The one benchmark that's noticeably slower is not a big deal, since that's the old format. We might look into the performance of it later but seems like good improvements across the board otherwise.

aryairani commented:
Should we update the benchmark to check the new serialization format? Or assume fine? (I would normally assume it's fine.)

dolio commented:
So, I've been looking at the optcheck failure. To me, it looks fishy.

When I compare the code it's complaining about to what I see when I enable the `+dumpcore` flag, they don't look similar. In the `+dumpcore` output `Stack` is completely optimized away. The only occurrences are in exception throws. In the `+optcheck` output, there are some join points that operate on `Stack`. They appear to be related to the `apply` function.

So, my conclusion is that the check isn't actually looking at fully optimized code, or the check itself is somehow changing what optimizations happen. I don't think it's the latter, because the check uses a plugin that just looks at GHC intermediate state. I guess in principle it _could_ influence the actual optimizer, because plugins can just change what compiler passes happen. But that's also unlikely.

I'm going to keep poking at things until the end of the day, but if you don't hear from me, I'd just turn off the check, and I'll file an issue with the library we use for this, asking for clarification. The _actual_ core looks fine, which is presumably why the benchmarks are fine. If the test isn't looking at the actual core, it's not a very useful test.

sellout committed:
Replace deprecated calls

This switches back to `-Werror=deprecations`, replacing all the
deprecated calls flagged in this upgrade.

sellout committed:
Disable `optchecks` in CI

sellout committed:
Run Weeder in CI with correct args

skykanin commented:
I can confirm that building and running unison from this PR works properly. It also fixes a weird linking issue I was running into on `trunk` while testing the new FFI additions.

From the unison discord:
> I'm getting an odd error when running my program with unison built from trunk (commit: 04f4bdaa3a1e27d94faf891309c9e2671dfcb474)
```
Failure
      (typeLink IOFailure)
      "user error (dlopen: /nix/store/aw2fw9ag10wr9pf0qk4nk5sxi0q0bn56-glibc-2.37-8/lib/libc.so.6: version `GLIBC_2.38' not found (required by /nix/store/7wpxzx3w1ci2izxmk6yd4zm5pir7z4wx-botan-3.10.0/lib/libbotan-3.so))"
      (Any ())

  > Stack trace:
    ##raise
```
> However when looking at the ldd output for the libbotan shared object it's clearly linked to a sufficiently new version of libc:
```
$ ldd /nix/store/7wpxzx3w1ci2izxmk6yd4zm5pir7z4wx-botan-3.10.0/lib/libbotan-3.so
    linux-vdso.so.1 (0x00007f9a38105000)
    libbz2.so.1 => /nix/store/bhz2i66mwm3vdyslixazss6a46h48ib5-bzip2-1.0.8/lib/libbz2.so.1 (0x00007f9a380e8000)
    libdl.so.2 => /nix/store/wb6rhpznjfczwlwx23zmdrrw74bayxw4-glibc-2.42-47/lib/libdl.so.2 (0x00007f9a380e3000)
    librt.so.1 => /nix/store/wb6rhpznjfczwlwx23zmdrrw74bayxw4-glibc-2.42-47/lib/librt.so.1 (0x00007f9a380de000)
    libz.so.1 => /nix/store/ri9paa3mri4kqakljak8ldvbcp7lpmif-zlib-1.3.1/lib/libz.so.1 (0x00007f9a380c0000)
    libstdc++.so.6 => /nix/store/j2kgllgds4w7na8zqv1msi0mpvpjxda8-gcc-15.2.0-lib/lib/libstdc++.so.6 (0x00007f9a37200000)
    libm.so.6 => /nix/store/wb6rhpznjfczwlwx23zmdrrw74bayxw4-glibc-2.42-47/lib/libm.so.6 (0x00007f9a37fc6000)
    libgcc_s.so.1 => /nix/store/j2kgllgds4w7na8zqv1msi0mpvpjxda8-gcc-15.2.0-lib/lib/libgcc_s.so.1 (0x00007f9a37f99000)
    libpthread.so.0 => /nix/store/wb6rhpznjfczwlwx23zmdrrw74bayxw4-glibc-2.42-47/lib/libpthread.so.0 (0x00007f9a37f94000)
    libc.so.6 => /nix/store/wb6rhpznjfczwlwx23zmdrrw74bayxw4-glibc-2.42-47/lib/libc.so.6 (0x00007f9a36e00000)
    /nix/store/wb6rhpznjfczwlwx23zmdrrw74bayxw4-glibc-2.42-47/lib64/ld-linux-x86-64.so.2 (0x00007f9a38107000)
```

sellout commented:
@aryairani This should be ready for review now.

dolio commented:
BTW, the response I received about the inspection testing thing was that it's actually possible for the mere presence of the inspection test to change the way GHC optimizes things. So that may be what we're seeing. But I'm not sure there's a concrete solution. The actual inspection test doesn't directly refer to the thing being optimized, so it's not anything as 'simple' as that.

aryairani commented:
I'm scared but here goes

mitchellwrosen commented:
@sellout I'm seeing a new error that you might know something about. Now when I run `git` in this repo (with config picked up by `use flake` in `.envrc`), I get:

```
warning: unhandled Platform key FamilyDisplayName
error: tool 'git' not found
```

I still see `git` on my path, it's at `/usr/bin/git`, but running it prints these lines for some reason. I'm on macOS (Intel). Any ideas?

aryairani commented:
@mitchellwrosen Oh, yeah I encountered this last night too, this line 50 here should fix it.

https://github.com/unisonweb/unison/pull/6150/changes/95a3ec805e9652b0146ef2f52a1a56ef8b19edc8#diff-8b67b4b0f775a1335829ea1852b4f6cd3a88983f051a2676f208ce6acd942c26

sellout commented:
@mitchellwrosen I haven’t seen that.

@aryairani For some reason, that link doesn’t show me anything helpful (or at least not obvious enough) – can you share the line contents?

I did a bit of poking around and noticed that the other message (warning: unhandled Platform key FamilyDisplayName) is maybe due to an interaction between some Nixpkgs 25.11 revisions and the Xcode-provided git on macOS (https://github.com/NixOS/nixpkgs/issues/376958) updating the flake (or your system’s Nixpkgs?) might get rid of that one.

aryairani commented:
@sellout @mitchellwrosen Sorry about that, here's the contents
<img width="801" height="235" alt="image" src="https://github.com/user-attachments/assets/5b7cc0a0-b0af-4432-aab8-8472fdad2ad2" />

> I did a bit of poking around and noticed that the other message (warning: unhandled Platform key FamilyDisplayName) is maybe due to an interaction between some Nixpkgs 25.11 revisions and the Xcode-provided git on macOS (https://github.com/NixOS/nixpkgs/issues/376958) updating the flake (or your system’s Nixpkgs?) might get rid of that one.

Maybe this would be better than adding it how I did (I should be able to use my system provided binaries too, IMO), but I'm not sure what it's suggesting to do about it.

mitchellwrosen commented:
Ah ha! Ok yeah, using git from nixpkgs 25.11 solved the problem. I think we want Arya's patch too so others don't hit this.

sellout commented:
@aryairani Weird – don’t know why my browser didn’t jump to that line.

Yeah, I do like keeping my flakes as self-contained as possible. For example, adding a `home-manager` input when I have `homeModules` or `homeConfigurations` in the flake and Nix even itself (although that one is a bit of a catch-22). `git` seems a bit more outside, but I think this issue justifies its place in a consistent environment (so yeah, keep @aryairani’s patch).

aryairani commented:
Ok cool. We can always pull it later if we figure out a better solution.
